compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/phases/common/WriteBarrierAdditionPhase.java

```diff
@@ -37,10 +37,16 @@
 
 public class WriteBarrierAdditionPhase extends BasePhase<CoreProviders> {
 
+    private final StageFlag stage;
+
+    public WriteBarrierAdditionPhase(StageFlag stage) {
+        this.stage = stage;
+    }
+
     @Override
     public Optional<NotApplicable> notApplicableTo(GraphState graphState) {
         return NotApplicable.ifAny(
-                        NotApplicable.ifApplied(this, StageFlag.BARRIER_ADDITION, graphState),
+                        NotApplicable.ifApplied(this, stage, graphState),
                         NotApplicable.unlessRunAfter(this, StageFlag.MID_TIER_LOWERING, graphState),
```

Comment:
This should probably be `stage == StageFlag.MID_TIER_BARRIER_ADDITION ? StageFlag.MID_TIER_LOWERING : StageFlag.FIXED_READS`.

Reply:
I've been thinking the same, but we don't always do FIXED_READS, right?

Reply:
True.  Then maybe just `LOW_TIER_LOWERING` then.  It's hard to get these right sometimes.